File: Programmers/Lv.2/괄호변환/괄호변환.java
package Programmers;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class 괄호변환 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String p = br.readLine();

        System.out.print(convert(p));
    }

    private static String convert(String p) {
        String result = "";
        int stk;
        char std;

        if (!isCorrect(p)) {
            std = p.charAt(0);
            stk = 1;
            String u = "", v = "";

            for (int i=1; i<p.length(); i++) {
                if (p.charAt(i) == std) {
                    stk += 1;
                } else {
                    stk -= 1;
                    if (stk == 0) {
                        u = p.substring(0, i+1);
                        v = p.substring(i+1);
                        break;
                    }
                }
            }

            if (isCorrect(u)) {
                result += u + convert(v);
            } else {
                result = "(" + convert(v) + ")";
                for (int i=1; i<u.length()-1; i++) {
                    if (u.charAt(i) == '(') {
                        result += ")";
                    } else {
                        result += "(";
                    }
                }
            }

        } else {
            result = p;
        }
        return result;
    }

    private static boolean isCorrect(String u) {
        if (u.length() > 0) {
            int stk = 0;
            for (int i=0; i<u.length(); i++) {
                if (u.charAt(i) == '(') 
                    stk += 1;
                else {
                    if (stk > 0) 
                        stk -= 1;
                    else 
                        return false;
                }
            }
            if (stk > 0)
                return false;
        } 
        return true;
    }
}

File: Programmers/Lv.2/괄호변환/괄호변환.py
def convert(p):    
    # 올바른 문자열이 아닐 때만
    if not isCorrect(p):
        u, v = '', ''
        std = ''
        result = ''
        std = p[0]
        count = 1
        # 앞에서부터 최소의 균형잡힌 문자열 찾기
        for i in range(1, len(p)):
            if p[i] == std:
                count += 1
            else:
                count -= 1
                if count == 0:
                    u = p[:i+1]
                    v = p[i+1:]
                    break
        
        if isCorrect(u):
            result += u + convert(v)
        else:
            result = '(' + convert(v) + ')'
            temp = u[1:len(u)-1]
            sub = ''
            for c in temp:
                if c == ')': 
                    sub += '('
                elif c == '(': sub += ')'   
            result += sub            
    
    else: 
        result = p  
    
    return result
    
def isCorrect(u):
    if len(u) > 0:
        stk = 0
        for c in u:
            if c == '(': 
                stk += 1
            elif c == ')' and stk > 0:
                stk -= 1
            else:
                return False
        if stk > 0:
            return False
    return True

def solution(p):
    answer = convert(p)
    return answer

def main():
    print(solution(input()))
    
if __name__ == "__main__":
    main()    